## Problem 2

# a
Code to find solution of $𝑥^{3} − 7𝑥^{2} + 14𝑥 − 5 = 0$ to a precision of $\epsilon < 10^{-8}$.

In [73]:
#BISECTION METHOD
#equation
def f(x):
    return x**3 - 7*x**2 + 14*x - 5

#initial guess
a = 0.0
b = 1.0
c_old = 1.0  # Placeholder to start
step = 0
rel_error = 1.0

while rel_error >= 1e-8:
    step += 1
    c = (a + b) / 2  # Find the midpoint
    
    # Calculate error (skipping the very first step)
    if step > 1:
        rel_error = abs((c - c_old) / c)
        
    print(f"Step {step}: x = {c:.8f}, Error = {rel_error:.8e}")
    
    # Update boundaries for the next loop
    if f(a) * f(c) < 0:
        b = c  # Root is in the left half
    else:
        a = c  # Root is in the right half
        
    c_old = c  # Save current midpoint for the next error calculation

print(f"\nFinal Answer: x = {c:.8f} found in {step} steps.")

Step 1: x = 0.50000000, Error = 1.00000000e+00
Step 2: x = 0.25000000, Error = 1.00000000e+00
Step 3: x = 0.37500000, Error = 3.33333333e-01
Step 4: x = 0.43750000, Error = 1.42857143e-01
Step 5: x = 0.46875000, Error = 6.66666667e-02
Step 6: x = 0.45312500, Error = 3.44827586e-02
Step 7: x = 0.46093750, Error = 1.69491525e-02
Step 8: x = 0.45703125, Error = 8.54700855e-03
Step 9: x = 0.45507812, Error = 4.29184549e-03
Step 10: x = 0.45410156, Error = 2.15053763e-03
Step 11: x = 0.45361328, Error = 1.07642626e-03
Step 12: x = 0.45336914, Error = 5.38502962e-04
Step 13: x = 0.45324707, Error = 2.69323997e-04
Step 14: x = 0.45318604, Error = 1.34680135e-04
Step 15: x = 0.45315552, Error = 6.73446023e-05
Step 16: x = 0.45317078, Error = 3.36711674e-05
Step 17: x = 0.45317841, Error = 1.68353003e-05
Step 18: x = 0.45318222, Error = 8.41757927e-06
Step 19: x = 0.45318031, Error = 4.20880735e-06
Step 20: x = 0.45318127, Error = 2.10439925e-06
Step 21: x = 0.45318174, Error = 1.05219852e-06
S

In [74]:
#NEWTON-RAPHSON METHOD
#equation
def f(x):
    return x**3 - 7*x**2 + 14*x - 5

def f_prime(x):
    # The derivative: 3x^2 - 14x + 14
    return 3*x**2 - 14*x + 14

x = 0.0  # initial guess
step = 0
rel_error = 1.0

while rel_error >= 1e-8:
    step += 1
    
    # Calculate the next guess using the Newton-Raphson formula
    x_new = x - (f(x) / f_prime(x))
    
    # Calculate the relative error
    rel_error = abs((x_new - x) / x_new)
        
    print(f"Step {step}: x = {x_new:.8f}, Error = {rel_error:.8e}")
    
    # Update our current guess for the next loop
    x = x_new

print(f"\nFinal Answer: x = {x:.8f} found in {step} steps.")

Step 1: x = 0.35714286, Error = 1.00000000e+00
Step 2: x = 0.44744815, Error = 2.01822917e-01
Step 3: x = 0.45315944, Error = 1.26032636e-02
Step 4: x = 0.45318172, Error = 4.91804013e-05
Step 5: x = 0.45318172, Error = 7.47456193e-10

Final Answer: x = 0.45318172 found in 5 steps.


Bisection method took 28 iterations and Newton-Raphson method took 5 iteration.

# b
Number of iterations to the location of initial guess. Here we are trying different initial guesses further or closer to the known 
answer 0.453.

In [75]:
#BISECTION METHOD
#equation
def f(x):
    return x**3 - 7*x**2 + 14*x - 5

#initial guess
a = -10
b = 10
c_old = 1.0  # Placeholder to start
step = 0
rel_error = 1.0

while rel_error >= 1e-8:
    step += 1
    c = (a + b) / 2  # Find the midpoint
    
    # Calculate error (skipping the very first step)
    if step > 1:
        rel_error = abs((c - c_old) / c)
        
    print(f"Step {step}: x = {c:.8f}, Error = {rel_error:.8e}")
    
    # Update boundaries for the next loop
    if f(a) * f(c) < 0:
        b = c  # Root is in the left half
    else:
        a = c  # Root is in the right half
        
    c_old = c  # Save current midpoint for the next error calculation

print(f"\nFinal Answer: x = {c:.8f} found in {step} steps.")

Step 1: x = 0.00000000, Error = 1.00000000e+00
Step 2: x = 5.00000000, Error = 1.00000000e+00
Step 3: x = 2.50000000, Error = 1.00000000e+00
Step 4: x = 1.25000000, Error = 1.00000000e+00
Step 5: x = 0.62500000, Error = 1.00000000e+00
Step 6: x = 0.31250000, Error = 1.00000000e+00
Step 7: x = 0.46875000, Error = 3.33333333e-01
Step 8: x = 0.39062500, Error = 2.00000000e-01
Step 9: x = 0.42968750, Error = 9.09090909e-02
Step 10: x = 0.44921875, Error = 4.34782609e-02
Step 11: x = 0.45898438, Error = 2.12765957e-02
Step 12: x = 0.45410156, Error = 1.07526882e-02
Step 13: x = 0.45166016, Error = 5.40540541e-03
Step 14: x = 0.45288086, Error = 2.69541779e-03
Step 15: x = 0.45349121, Error = 1.34589502e-03
Step 16: x = 0.45318604, Error = 6.73400673e-04
Step 17: x = 0.45303345, Error = 3.36813742e-04
Step 18: x = 0.45310974, Error = 1.68378515e-04
Step 19: x = 0.45314789, Error = 8.41821702e-05
Step 20: x = 0.45316696, Error = 4.20893135e-05
Step 21: x = 0.45317650, Error = 2.10442139e-05
S

In [76]:
#NEWTON-RAPHSON METHOD
#equation
def f(x):
    return x**3 - 7*x**2 + 14*x - 5

def f_prime(x):
    # The derivative: 3x^2 - 14x + 14
    return 3*x**2 - 14*x + 14

x = 1.7  # initial guess
step = 0
rel_error = 1.0

while rel_error >= 1e-8:
    step += 1
    
    # Calculate the next guess using the Newton-Raphson formula
    x_new = x - (f(x) / f_prime(x))
    
    # Calculate the relative error
    rel_error = abs((x_new - x) / x_new)
        
    print(f"Step {step}: x = {x_new:.8f}, Error = {rel_error:.8e}")
    
    # Update our current guess for the next loop
    x = x_new

    
print(f"\nFinal Answer: x = {x:.8f} found in {step} steps.")

Step 1: x = 4.78230088, Error = 6.44522576e-01
Step 2: x = 4.06497872, Error = 1.76463941e-01
Step 3: x = 3.55296650, Error = 1.44108374e-01
Step 4: x = 2.97637927, Error = 1.93721019e-01
Step 5: x = 3.91407592, Error = 2.39570378e-01
Step 6: x = 3.42583637, Error = 1.42516891e-01
Step 7: x = 2.61287108, Error = 3.11138693e-01
Step 8: x = 3.38891604, Error = 2.28995040e-01
Step 9: x = 2.42558039, Error = 3.97156763e-01
Step 10: x = 3.31161769, Error = 2.67554224e-01
Step 11: x = 1.61419388, Error = 1.05156129e+00
Step 12: x = 6.17433465, Error = 7.38563914e-01
Step 13: x = 4.98262330, Error = 2.39173478e-01
Step 14: x = 4.19896960, Error = 1.86630000e-01
Step 15: x = 3.65637118, Error = 1.48398070e-01
Step 16: x = 3.14640601, Error = 1.62078628e-01
Step 17: x = 5.71616905, Error = 4.49560364e-01
Step 18: x = 4.68240632, Error = 2.20775955e-01
Step 19: x = 3.99746543, Error = 1.71343792e-01
Step 20: x = 3.49788589, Error = 1.42823282e-01
Step 21: x = 2.85167423, Error = 2.26607813e-01
S

Bisection method is very insenitive while Newton-Rapson method is highly sensitive to the initial guess.

# c
Initial guess for code to go astray

In [77]:
import numpy as np
#equation
def f(x):
    return x**3 - 7*x**2 + 14*x - 5

def f_prime(x):
    # The derivative: 3x^2 - 14x + 14
    return 3*x**2 - 14*x + 14

x = (7 - np.sqrt(7)) / 3  # initial guess
step = 0
rel_error = 1.0

while rel_error >= 1e-8:
    step += 1
    
    # Calculate the next guess using the Newton-Raphson formula
    x_new = x - (f(x) / f_prime(x))
    
    # Calculate the relative error
    rel_error = abs((x_new - x) / x_new)
        
    print(f"Step {step}: x = {x_new:.8f}, Error = {rel_error:.8e}")
    
    # Update our current guess for the next loop
    x = x_new

    
print(f"\nFinal Answer: x = {x:.8f} found in {step} steps.")

Step 1: x = -inf, Error = nan

Final Answer: x = -inf found in 1 steps.


C:\Users\allst\AppData\Local\Temp\ipykernel_16468\1582302607.py:18: RuntimeWarning: divide by zero encountered in scalar divide
  x_new = x - (f(x) / f_prime(x))
C:\Users\allst\AppData\Local\Temp\ipykernel_16468\1582302607.py:21: RuntimeWarning: invalid value encountered in scalar divide
  rel_error = abs((x_new - x) / x_new)


For those values of initial guesses where derivative of function is zero, the code will go astray. For example if the initial guess $x_0 = (7 - \sqrt 7)/3$, where the derivative of the given function is zero, the code won't work. For the initial guesses, that are close to roots of derivative of the function, number of iteration  will be very high to get to the precision of $\epsilon < 10^{-8}$. For example if the initial guess $x_0 = 1.45141623, it will take 74 iteration.